core/src/server/helpers.rs

```diff
@@ -196,8 +198,41 @@ pub async fn collect_batch_response(rx: mpsc::UnboundedReceiver<String>) -> Stri
 	buf
 }
 
+/// Wrapper over [`tokio::sync::Notify`] with bounds check.
+#[derive(Debug)]
+pub struct BoundedSubscriptions {
```

Comment:
I was just sortof wodnering whether this could be replaced with a Semaphore (https://docs.rs/tokio/latest/tokio/sync/struct.Semaphore.html), which would start at `max_subscriptions` and hand out permits for subscriptions until it runs out (and starts returning errors).

Reply:
yeah, that would be nice the hard to part is where to store the semaphore after acquiring it.

I guess we could move it to the callbacks but that's quite annoying.

Reply:
I'm happy to not worry about it for now; we can always have a pass later to see whether we can simplify/optimise things!